tools-local/Microsoft.ML.InternalCodeAnalyzer/BaseTestClassAnalyzer.cs

```diff
@@ -0,0 +1,134 @@
+        {
+            if (!(context.Compilation.GetTypeByMetadataName("Xunit.FactAttribute") is { } factAttribute))
+            {
+                return;
+            }
+
+            var analyzerImpl = new AnalyzerImpl(context.Compilation, factAttribute);
+            context.RegisterSymbolAction(analyzerImpl.AnalyzeNamedType, SymbolKind.NamedType);
+        }
+
+        private sealed class AnalyzerImpl
+        {
+            private readonly Compilation _compilation;
+            private readonly INamedTypeSymbol _factAttribute;
+            private readonly INamedTypeSymbol _baseTestClass;
+            private readonly ConcurrentDictionary<INamedTypeSymbol, bool> _knownTestAttributes = new ConcurrentDictionary<INamedTypeSymbol, bool>();
+
+            public AnalyzerImpl(Compilation compilation, INamedTypeSymbol factAttribute)
+            {
+                _compilation = compilation;
+                _factAttribute = factAttribute;
+                _baseTestClass = _compilation.GetTypeByMetadataName("Microsoft.ML.TestFramework.BaseTestClass");
+            }
+
+            public void AnalyzeNamedType(SymbolAnalysisContext context)
+            {
+                var namedType = (INamedTypeSymbol)context.Symbol;
+                if (namedType.TypeKind != TypeKind.Class)
+                    return;
+
+                if (ExtendsBaseTestClass(namedType))
+                    return;
+
+                var hasTestMethod = false;
+                foreach (var member in namedType.GetMembers())
+                {
+                    if (member is IMethodSymbol method && IsTestMethod(method))
+                    {
+                        hasTestMethod = true;
+                        break;
+                    }
+                }
+
+                if (!hasTestMethod)
+                    return;
+
+                context.ReportDiagnostic(Diagnostic.Create(Rule, namedType.Locations[0], namedType));
+            }
+
+            private bool IsTestMethod(IMethodSymbol method)
+            {
+                foreach (var attribute in method.GetAttributes())
+                {
+                    if (IsTestAttribute(attribute.AttributeClass))
+                        return true;
+                }
+
+                return false;
+            }
+
+            private bool IsTestAttribute(INamedTypeSymbol attributeClass)
+            {
+                if (_knownTestAttributes.TryGetValue(attributeClass, out var isTest))
+                    return isTest;
+
+                return _knownTestAttributes.GetOrAdd(attributeClass, ExtendsFactAttribute(attributeClass));
+            }
+
+            private bool ExtendsBaseTestClass(INamedTypeSymbol namedType)
+            {
+                if (_baseTestClass is null)
+                    return false;
+
+                for (var current = namedType; current is object; current = current.BaseType)
+                {
+                    if (Equals(current, _baseTestClass))
```

Comment:
Symbols need to be compared with `Equals` (as opposed to `==`). We have an analyzer for this to help developers avoid mistakes.